structuretoolkit/analyse/snap.py

```diff
@@ -0,0 +1,397 @@
+
+    return np.array(((xhi, 0, 0), (xyp, yhi, 0), (xzp, yzp, zhi)))
+
+
+def write_ase_structure(lmp, structure):
+    number_species = len(set(structure.get_chemical_symbols()))
+
+    apre = get_apre(cell=structure.cell)
+    (
+        (xhi, xy, xz),
+        (_, yhi, yz),
+        (_, _, zhi),
+    ) = apre.T
+    lmp.command(
+        "region 1 prism"
+        + " 0.0 "
+        + str(xhi)
+        + " 0.0 "
+        + str(yhi)
+        + " 0.0 "
+        + str(zhi)
+        + " "
+        + str(xy)
+        + " "
+        + str(xz)
+        + " "
+        + str(yz)
+        + " units box"
+    )
+    lmp.command("create_box " + str(number_species) + " 1")
+
+    el_dict = {el: i for i, el in enumerate(set(structure.get_chemical_symbols()))}
+
+    R = np.dot(np.linalg.inv(structure.cell), apre)
+
+    positions = structure.positions.flatten()
+    if np.matrix.trace(R) != 3:
+        positions = np.array(positions).reshape(-1, 3)
+        positions = np.matmul(positions, R)
+    positions = positions.flatten()
+    elem_all = np.array([el_dict[el] + 1 for el in structure.get_chemical_symbols()])
+    lmp.create_atoms(
+        n=len(structure),
+        id=None,
+        type=(len(elem_all) * c_int)(*elem_all),
+        x=(len(positions) * c_double)(*positions),
+        v=None,
+        image=None,
+        shrinkexceed=False,
+    )
+
+
+def extract_compute_np(lmp, name, compute_type, result_type, array_shape):
+    """
+    Convert a lammps compute to a numpy array.
+    Assumes the compute returns a floating point numbers.
+    Note that the result is a view into the original memory.
+    If the result type is 0 (scalar) then conversion to numpy is skipped and a python float is returned.
+    """
+    ptr = lmp.extract_compute(
+        name, compute_type, result_type
+    )  # 1,2: Style (1) is per-atom compute, returns array type (2).
+    if result_type == 0:
+        return ptr  # No casting needed, lammps.py already works
+    if result_type == 2:
+        ptr = ptr.contents
+    total_size = int(np.prod(array_shape))
+    buffer_ptr = cast(ptr, POINTER(c_double * total_size))
+    array_np = np.frombuffer(buffer_ptr.contents, dtype=float)
+    array_np.shape = array_shape
+    return array_np.copy()
+
+
+def reset_lmp(lmp):
+    lmp.command("clear")
+    lmp.command("units metal")
+    lmp.command("dimension 3")
+    lmp.command("boundary p p p")
+    lmp.command("atom_style charge")
+    lmp.command("atom_modify map array sort 0 2.0")
+
+
+def set_potential_lmp(lmp, cutoff=10.0):
+    lmp.command("pair_style zero " + str(cutoff))
+    lmp.command("pair_coeff * *")
+    lmp.command("mass * 1.0e-20")
+    lmp.command("neighbor 1.0e-20 nsq")
+    lmp.command("neigh_modify one 10000")
+
+
+def get_snap_descriptor_names(twojmax):
+    lst = []
+    for j1 in range(0, twojmax + 1):
+        for j2 in range(0, j1 + 1):
+            for j in range(j1 - j2, min(twojmax, j1 + j2) + 1, 2):
+                if j >= j1:
```

Comment:
This is unnecessary as the entire innermost loop could instead just be truncated to start at `j1`

Reply:
The increment is `2` rather than `1`, so just starting at `j1` is not sufficient.

Reply:
So you suggest to move the if statement one level up? I am not convinced that his improves the readability of the code.

Reply:
I'm saying replace j1-j2 and the if with a simple start=j2%2. Don't loop over stuff you're going to skip. This isn't about readability, it's about shortening the for loop.

Reply:
I'm wrong about the nod, yes, moving the if our should do it

Reply:
Actually it's probably both, pull up j2<j1 as the if, then start at j2%2. But I'm on mobile and need to get some kids breakfast so I'll have to leave the details to you. No wasted loop space please though

Reply:
Done